src/primitives/threshold.rs

```diff
@@ -0,0 +1,310 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Thresholds
+//!
+//! Miniscript
+
+#[cfg(all(not(feature = "std"), not(test)))]
+use alloc::vec;
+use core::{cmp, fmt, iter};
+#[cfg(any(feature = "std", test))]
+use std::vec;
+
+/// Error parsing an absolute locktime.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ThresholdError {
+    k: usize,
+    n: usize,
+    max: Option<usize>,
+}
+
+impl fmt::Display for ThresholdError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.n == 0 {
+            f.write_str("thresholds in Miniscript must be nonempty")
+        } else if self.k == 0 {
+            f.write_str("thresholds in Miniscript must have k > 0")
+        } else if self.k > self.n {
+            write!(f, "invalid threshold {}-of-{}; cannot have k > n", self.k, self.n)
+        } else {
+            debug_assert!(self.max.is_some());
+            let max = self.max.unwrap();
+            debug_assert!(self.n > max);
+            write!(f, "invalid threshold {}-of-{}; maximum size is {}", self.k, self.n, max)
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for ThresholdError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
+}
+
+/// Structure representing a k-of-n threshold collection of some arbitrary
+/// object `T`.
+///
+/// If the constant parameter `MAX` is nonzero, it represents a cap on the
+/// `n` value; if `n` exceeds `MAX` then an error is returned on construction.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Threshold<T, const MAX: usize = 0> {
+    k: usize,
+    inner: Vec<T>,
```

Comment:
Slight preference for `v` (or something else) because this is not a wrapper type so `inner` seems slightly off.